src/MizarData/include/MizarData/MizarPairedData.h

```diff
@@ -198,7 +203,45 @@ class MizarPairedDataTmpl {
   absl::flat_hash_map<TID, uint64_t> tid_to_callstack_samples_counts_;
 };
 
-using MizarPairedData = MizarPairedDataTmpl<MizarDataProvider, FrameTrackManager>;
+template <typename Data, auto GetSamplingIntervalsNs>
+uint64_t RobustSamplingPeriodEstimate(const absl::flat_hash_set<orbit_mizar_base::TID>& tids,
+                                      uint64_t min_timestamp, uint64_t max_timestamp,
+                                      const Data& data) {
+  std::vector<uint64_t> intervals = GetSamplingIntervalsNs(
+      tids, min_timestamp, max_timestamp, data.GetCaptureData().GetCallstackData());
+  absl::c_sort(intervals);
+
+  constexpr double kOutlierRate = 0.025;
+  const size_t cutoff_index = static_cast<size_t>((1 - kOutlierRate) * intervals.size());
+
+  if (cutoff_index + 1 > intervals.size()) {
+    return data.GetNominalSamplingPeriodNs();
+  }
+  const uint64_t cutoff = intervals[cutoff_index];
+
+  struct TotalAndCount {
+    uint64_t total{};
+    size_t count{};
+
+    // round(total / count)
+    uint64_t ComputeAverage() { return (total + (count / 2)) / count; }
+  };
+
+  return absl::c_accumulate(
```

Comment:
I believe this construct (the accumulate and the struct) is not really helping readability. I would highly recommend something like this:

```
uint64_t total = 0;
size_t count = 0;

for (uint64_t interval : intervals) {
  if (interval > cuttoff) continue;
  total += interval;
  ++count;
}
return total + (count / 2) / count;
```

It's not even shorter, but you also know directly what is happening.

Reply:
OK